### PR TITLE
Update image custom plugin

### PR DESCRIPTION
**Pull request by Bondok6**
Updated our custom image plugin

- I removed the old `image` plugin `helpjuice-build-v2/src/plugins/image`
- I copied the original plugin `ckeditor/ckeditor5-image` and put it in place of the old one `helpjuice-build-v2/src/plugins/image`
- Then I added these changes https://github.com/Helpjuice/ckeditor5/commit/3529f554bf6e52ae0a5865a6f286bcb58adc3841

I also installed a package `fork-ts-checker-wepackk-plugin` to speed up the build process and it's useful for debugging ( gives us better error reporting in the console )

[helpjuice4 repo PR](https://github.com/hajrice/helpjuice4/pull/3062)

### CODE REVIEW

**Review by NicholasJarr — COMMENTED** [review]
@Bondok6

I checked and most of the changes are for adding Typescript types or something similar. Before moving forward with it, are the only differences between this plugin and the original one the changes on this commit here?

https://github.com/Helpjuice/ckeditor5/commit/3529f554bf6e52ae0a5865a6f286bcb58adc3841

If that's the case, I think we can extend the schema without having to copy everything from the original plugin. Can you check if that's possible? If it is, it can make things simpler both for this update and future ones.

**Comment by Bondok6** [comment]
> 



> @Bondok6
> 
> I checked and most of the changes are for adding Typescript types or something similar. Before moving forward with it, are the only differences between this plugin and the original one the changes on this commit here?
> 
> [3529f55](https://github.com/Helpjuice/ckeditor5/commit/3529f554bf6e52ae0a5865a6f286bcb58adc3841)

I compared the content manually, and I think so.
I also tested the image feature on the staging and it works without issues

> If that's the case, I think we can extend the schema without having to copy everything from the original plugin. Can you check if that's possible? If it is, it can make things simpler both for this update and future ones.

@NicholasJarr 
Sorry, I didn't get that! Can you explain what you mean by extending the schema?

**Comment by NicholasJarr** [comment]
> @NicholasJarr Sorry, I didn't get that! Can you explain what you mean by extending the schema?

My bad. I meant using the [`Schema#extend`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_model_schema-Schema.html#function-extend) method to add new rules to an existing model in CKEditor. In the original commit, there's this change here:

image

We could register a new plugin (that has `ImageInline` and `ImageEditing` in `requires`) that executes the following code on `init`:

```js
schema.extend( 'imageInline', {
  allowAttributes: [ 'alt', 'src', 'srcset','width', 'height', 'style' ]
});
```

We could do the same for these changes too:

image

After the new plugin is set up, we can register it in the editor and remove our custom Image plugin. Any thoughts?

File: packages/ckeditor5-build-helpjuice-v2/src/plugins/image/customimage.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import ImageInline from '@ckeditor/ckeditor5-image/src/imageinline';
+import ImageBlock from '@ckeditor/ckeditor5-image/src/imageblock';
+
+export default class MyImagePlugin extends Plugin {
```
Comment (NicholasJarr) [inline packages/ckeditor5-build-helpjuice-v2/src/plugins/image/customimage.js]: Can you change the name to something different? Maybe `CustomImageAttributesPlugin` (in this case you can also change the filename)?

**Review by NicholasJarr — COMMENTED** [review]
Awesome! This can really help reducing the amount of code to maintain in the editor. Just two points besides the comment I made:

1. When I built locally with `npm build`, it changed two files `package-lock.json` and `hjeditor.js.map`. Is the latest build on the branch?
2. I copied the built file to my local app and I tested it out. The width worked without problems, but the height doesn't work. Any thoughts?

File: packages/ckeditor5-build-helpjuice-v2/src/plugins/image/customimage.js (hunk 1 above)
Comment (Bondok6) [reply packages/ckeditor5-build-helpjuice-v2/src/plugins/image/customimage.js]: Done

**Comment by Bondok6** [comment]
> Awesome! This can really help reducing the amount of code to maintain in the editor. Just two points besides the comment I made:
> 
> 1. When I built locally with `npm build`, it changed two files `package-lock.json` and `hjeditor.js.map`. Is the latest build on the branch?

yes the latest build on the branch

BTW, I made one change to the package.json, I added a new script `"dev": "webpack --mode development --watch"`

```
  "scripts": {
     .....
    "dev": "webpack --mode development --watch"
  }
  ```
  
  Not sure if it could be the reason.

> 2. I copied the built file to my local app and I tested it out. The width worked without problems, but the height doesn't work. Any thoughts?

It works on my end, Can you test it again? And how did you test it maybe I'm missing something?

**Comment by NicholasJarr** [comment]
> Not sure if it could be the reason.

It's a change to the `fork-ts-checker-wepackk-plugin`, but it may be something on my environment.

> It works on my end, Can you test it again? And how did you test it maybe I'm missing something?

Still doesn't work for me. I copied `hjeditor.js` from this branch and pasted it on my local app repo. Then, I restarted my Rails server and webpack server. Then, I inserted an image using the Files Manager. The rest of the steps you can see in the video below:


https://github.com/Helpjuice/ckeditor5/assets/28270050/26ef8562-cc3e-408a-931d-294c7a35f544

**Comment by Bondok6** [comment]
> > Not sure if it could be the reason.
> 
> It's a change to the `fork-ts-checker-wepackk-plugin`, but it may be something on my environment.
> 
> > It works on my end, Can you test it again? And how did you test it maybe I'm missing something?
> 
> Still doesn't work for me. I copied `hjeditor.js` from this branch and pasted it on my local app repo. Then, I restarted my Rails server and webpack server. Then, I inserted an image using the Files Manager. The rest of the steps you can see in the video below:
> 
>  2023-07-13.18-27-46.mp4

@NicholasJarr, 

The problem is in the original plugin `imageinlineediting.ts` and `imageinlineediting.ts`
There is a `private function` called `_setupConversion()` that contains

```ts

		conversion.for( 'downcast' )
			.add( downcastImageAttribute( imageUtils, 'imageInline', 'src' ) )
			.add( downcastImageAttribute( imageUtils, 'imageInline', 'alt' ) )
			.add( downcastSrcsetAttribute( imageUtils, 'imageInline' ) );
```

Adding the width and height here will fix the problem

```the
			.add( downcastImageAttribute( imageUtils, 'imageInline', 'width' ) )
			.add( downcastImageAttribute( imageUtils, 'imageInline', 'height' ) )
```

But I'm not able to modify this private function in our Custom plugin.

So there are two solutions

1) To modify the original plugin. However, this is not recommended as it can lead to compatibility issues and make it difficult to update the plugin in the future.

2) To copy the original `image` plugin and update it, I didn't copy everything I just copied the `imageinline` and `imageblock` and modified them. (That's what I did)

in `hjeditor.ts` we're still importing everything from the original plugin, instead, the image plugin that contains ( `imageinline` and `imageblock` )

```ts
// Image Plugin
import { ImageCaption, ImageStyle, ImageResize, ImageToolbar, ImageUpload } from '@ckeditor/ckeditor5-image';
import CustomImage from './plugins/image/image';
```


> The issue should be fixed 

[Screencast from 16 يول, 2023 EEST 06:40:41 ص.webm](https://github.com/Helpjuice/ckeditor5/assets/45231276/7c98941f-3569-4ff3-99a6-8a535979045e)

**Review by NicholasJarr — APPROVED** [review]

**Comment by NicholasJarr** [comment]
Hey @Bondok6!

Nice work! I tested and it worked without issues. This should make it easier to maintain the image plugin on our next updates. I think we can still investigate this and possibly fix the bug without having to copy code from the original editor, but it may take longer to investigate and we need to have the editor updated for deploying the other fixes. Can you open a card for this?

**Comment by Bondok6** [comment]
> Hey @Bondok6!
> 
> Nice work! I tested and it worked without issues. This should make it easier to maintain the image plugin on our next updates. I think we can still investigate this and possibly fix the bug without having to copy code from the original editor, but it may take longer to investigate and we need to have the editor updated for deploying the other fixes. Can you open a card for this?

Hey @NicholasJarr, I 100% agree with you will open a card for it

[Card](https://app.shortcut.com/helpjuice/story/3960/ckeditor-improve-our-image-plugin-try-to-not-have-copied-from-the-original-image-plugin)

**Pull request merged**